fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/SkewShuffleJoinEliminationRule.java

```diff
@@ -0,0 +1,514 @@
+                            DistributionSpec.createReplicatedDistributionSpec(), rightSplitOutputColumnRefMap);
+
+            Projection projectionOnJoin = originalShuffleJoinOperator.getProjection();
+
+            PhysicalHashJoinOperator newShuffleJoinOpt = new PhysicalHashJoinOperator(
+                    originalShuffleJoinOperator.getJoinType(), originalShuffleJoinOperator.getOnPredicate(),
+                    originalShuffleJoinOperator.getJoinHint(), originalShuffleJoinOperator.getLimit(),
+                    originalShuffleJoinOperator.getPredicate(), projectionOnJoin,
+                    originalShuffleJoinOperator.getSkewColumn(), originalShuffleJoinOperator.getSkewValues());
+
+            PhysicalHashJoinOperator newBroadcastJoinOpt = new PhysicalHashJoinOperator(
+                    originalShuffleJoinOperator.getJoinType(), originalShuffleJoinOperator.getOnPredicate(),
+                    originalShuffleJoinOperator.getJoinHint(), originalShuffleJoinOperator.getLimit(),
+                    originalShuffleJoinOperator.getPredicate(), projectionOnJoin,
+                    originalShuffleJoinOperator.getSkewColumn(), originalShuffleJoinOperator.getSkewValues());
+
+            LocalExchangerType localExchangerType =
+                    parentRequireEmpty ? LocalExchangerType.DIRECT : LocalExchangerType.PASS_THROUGH;
+            PhysicalMergeOperator mergeOperator =
+                    buildMergeOperator(opt.getOutputColumns().getColumnRefOperators(columnRefFactory), 2,
+                            localExchangerType, originalShuffleJoinOperator.getLimit());
+
+            OptExpression leftSplitConsumerOptExpForShuffleJoin = OptExpression.builder()
+                    .setOp(leftSplitConsumerOptForShuffleJoin)
+                    .setInputs(Collections.emptyList())
+                    .setLogicalProperty(leftSplitProduceOptExp.getLogicalProperty())
+                    .setStatistics(leftSplitProduceOptExp.getStatistics())
+                    .setCost(leftSplitProduceOptExp.getCost()).build();
+
+            OptExpression rightSplitConsumerOptExpForShuffleJoin = OptExpression.builder()
+                    .setOp(rightSplitConsumerOptForShuffleJoin)
+                    .setInputs(Collections.emptyList())
+                    .setLogicalProperty(rightSplitProduceOptExp.getLogicalProperty())
+                    .setStatistics(rightSplitProduceOptExp.getStatistics())
+                    .setCost(rightSplitProduceOptExp.getCost()).build();
+
+            OptExpression newShuffleJoin = OptExpression.builder()
+                    .setOp(newShuffleJoinOpt)
+                    .setInputs(List.of(leftSplitConsumerOptExpForShuffleJoin, rightSplitConsumerOptExpForShuffleJoin))
+                    .setLogicalProperty(opt.getLogicalProperty())
+                    .setStatistics(opt.getStatistics())
+                    .setRequiredProperties(opt.getRequiredProperties())
```

Comment:
Is the new join OptExpression requiredProperties for its children same with the old join OptExpression? I think it's different.

Reply:
> Is the new join OptExpression requiredProperties for its children same with the old join OptExpression? I think it's different.
I think new shuffle join is the same as old shuffle join, broadcast join is different :
```
OptExpression newBroadcastJoin = OptExpression.builder().setOp(newBroadcastJoinOpt)
                    .setInputs(
                            List.of(leftSplitConsumerOptExpForBroadcastJoin, rightSplitConsumerOptExpForBroadcastJoin))
                    .setLogicalProperty(opt.getLogicalProperty())
                    .setStatistics(opt.getStatistics())
                    .setRequiredProperties(requiredPropertiesForBroadcastJoin)
                    .setCost(opt.getCost()).build();
```